dev/core#1036 Move code from Tell-a-friend to a core extension

mlutfy opened the pull request:
Overview
----------------------------------------

Towards deprecating Tell-a-friend: https://lab.civicrm.org/dev/core/-/issues/1036

Moves part of the code into a core extension, which is not enabled by default on new installs, but is enabled on existing installs (as part of the upgrader).

Before
----------------------------------------

The "Tell a friend" tab is always visible on Manage Contribution Page, and Manage Event. 

After
----------------------------------------

The tab is not visible on new installs. For existing installs, it is enabled only if it is used.

A status check warning is displayed to admins:

![image](https://github.com/civicrm/civicrm-core/assets/254741/7adddfee-4db5-4ea2-a0db-76d62f42586c)


Technical Details
----------------------------------------

~~There is a bit about AllCoreTables that I don't know what the implications are, but it would crash CiviCRM if I didn't remove that code.~~ (the file has been removed recently)

Comments
----------------------------------------

There are bits of code still not moved, such as:

- the Menu xml
- database table creation

I would not aim to have it perfect. Having bits in the core extension moves a lot out of core, but if we want to completely make the database tables optional, then there is a lot of old code that needs to be wrapped in conditions in order to make it work.

We could wait a year, and based on the feedback of the experiment, invest more time in restructuring code, or just remove the code completely? (including the core ext)

mlutfy committed:
dev/core#1036 Move bits from Tell-a-friend to a core extension

mlutfy commented:
cc @herbdool @MegaphoneJon (since you had expressed interest in this)

mlutfy commented on CRM/Core/DAO/AllCoreTables.data.php:
```diff
@@ -227,11 +227,6 @@
-    'name' => 'Friend',
-    'class' => 'CRM_Friend_DAO_Friend',
-    'table' => 'civicrm_tell_friend',
-  ],
```
What does this do?

mlutfy replied:
File was removed upstream with the DAO removals, which answers my question :)

herbdool commented:
I'll take a look this week. But, first, isn't it a long weekend for you too? 😉 And why am I looking at my emails on this sunny day?

colemanw commented on CRM/Event/Form/ManageEvent/TabHeader.php:
```diff
@@ -85,7 +85,9 @@ public static function process(&$form) {
-    $tabs['friend'] = ['title' => ts('Tell a Friend')] + $default;
+    if (function_exists('tellafriend_config')) {
+      $tabs['friend'] = ['title' => ts('Tell a Friend')] + $default;
+    }
```
I'd think this could be done with `hook_civicrm_tabSet()`

mlutfy replied:
That specific one is a deprecated function scheduled for deletion in 5.78, but I fixed the other place for Events.

colemanw commented on CRM/Upgrade/Incremental/php/FiveSeventyFive.php:
```diff
@@ -71,6 +71,40 @@ public function upgrade_5_75_alpha1($rev): void {
+  public static function installTellafriend(CRM_Queue_TaskContext $ctx) {
+    // Based on the instructions for the FiveThirty financialacls upgrade step
+    // Install via direct SQL manipulation. Note that:
+    // (1) This extension has no activation logic as of 5.75 (the DB tables are still in core)
```
If the DB tables are still in core, then the xml/schema files should stay there too.

colemanw commented on CRM/Contribute/Form/ContributionPage/TabHeader.php:
```diff
@@ -94,6 +94,10 @@ public static function process(&$form) {
+
     $contribPageId = $form->getVar('_id');
     // Call tabset hook to add/remove custom tabs
     CRM_Utils_Hook::tabset('civicrm/admin/contribute', $tabs, ['contribution_page_id' => $contribPageId]);
```
This hook can be used by the extension & the above code can be removed from core.

mlutfy replied:
Done. It moves the "Tell a Friend" to the last tab, but I think that's OK, since we're deprecating it.

mlutfy commented:
@colemanw Cool thanks. I'll move the XML files back, and will look into the tabset hook.

Question: does it make sense to leave the extension disabled by default, and enable it for when the tests run? Is there a recommended way of enabling it for tests? I would rather leave the tests in core as they are, at least for now.

mlutfy commented:
Some of the test errors:

```
1) E2E_Shimmy_LifecycleTest::testLifecycleWithSubprocesses
RuntimeException: Command failed (cd '/home/homer/buildkit/build/build-2/web/sites/default/files/civicrm/ext/example-mixin'; cv api3 --in=json 'Extension.disable'):

/home/homer/buildkit/build/build-2/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:186
/home/homer/buildkit/build/build-2/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:136
/home/homer/buildkit/build/build-2/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:68
/home/homer/buildkit/build/build-2/web/sites/default/files/civicrm/ext/example-mixin/tests/phpunit/E2E/Shimmy/LifecycleTest.php:43
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2307

2) E2E_Shimmy_LifecycleTest::testLifecycleWithLocalFunctions
Error: Class "CRM_Friend_DAO_Friend" not found
```

mlutfy commented:
I did a few important fixes, implemented `civicrm_hook_tabset` and also added a Status Check warning.

mlutfy commented:
I also removed the stuff from the "generated" data, so that it's not set by default on sandboxes (not that it has a huge impact, because the extension itself won't be enabled).

mlutfy commented:
Bug: on the sandbox site, this shows up even if the extension is disabled: 

![image](https://github.com/civicrm/civicrm-core/assets/254741/5457ccac-d791-4d8f-9865-131c5900ef17)

Rest seems mostly OK, except for the many test fails, some of which are rather confusing, but I'll wait for the report to finish.

mlutfy commented:
yay  tests :tada:

mlutfy commented:
I pushed small changes:

- Removed the "Tell a friend" menu from the old Manage Contributions page (was in a screenshot in a previous comment). Previously the link was always shown. I didn't bother adding it back via the "tabset" hook because that hook behaves pretty horribly depending on the context, and it was causing some mysterious fatal errors. I don't think anyone will really be bitter they now have to do 2 clicks for that feature.
- Moved the xml menu entry from core to the extension.

So besides the DAOs still in core, I think this is ready for review, and hopefully to merge. The tests are not complaining about the extension being disabled, because there were no tests (only some symbolic tests on the DB record creation, it seems).

In this experience, I have also learned how frustrating it can be to untangle and move core code to an extension, for very little functional benefit and no pay, other than having cleaner code (which, as a dev, I greatly appreciate), and probably make a few people angry about the change (been there too). Hat-tip to those who have been doing those cleanups on a regular basic.

herbdool commented on CRM/Contribute/BAO/ContributionPage.php:
```diff
@@ -742,11 +742,15 @@ public static function getSectionInfo($contribPageIds = []) {
       'premium',
     ];
+
+    if (function_exists('tellafriend_civicrm_config')) {
```
```suggestion
    if (\Civi\Api4\Extension::get(FALSE)->addWhere('file', '=', 'tellafriend')->addWhere('status', '=', 'installed')->execute()->selectRowCount()->count()) {
```

It seems a bit more legit to use Api4. Maybe there's a shorter way to write this?

demeritcowboy replied:
Zooming out, I think the question to ask is "how should extensions modify core", to which the answer is "implement hooks!" I haven't looked too close but for this specific spot maybe hook_tabset, or buildForm, can do it. But then zooming sideways, is that amount of refactoring something you want to take on in this PR.

mattwire replied:
For now I think this is fine. Further refactoring is much easier once this PR is merged. Feel free to follow up with this change or an alternative using hooks.

herbdool commented on CRM/Contribute/Form/Contribution/ThankYou.php:
```diff
@@ -245,45 +245,46 @@ public function buildQuickForm() {
+    $friendURL = NULL;
+    $friendText = NULL;
+
+    if (function_exists('tellafriend_civicrm_config')) {
```
```suggestion
    if (\Civi\Api4\Extension::get(FALSE)->addWhere('file', '=', 'tellafriend')->addWhere('status', '=', 'installed')->execute()->selectRowCount()->count()) {
```

herbdool commented on CRM/Contribute/StateMachine/ContributionPage.php:
```diff
@@ -46,6 +46,9 @@ public function __construct($controller, $action = CRM_Core_Action::NONE) {
       'CRM_Contribute_Form_ContributionPage_Widget' => NULL,
     ];
 
+    if (!function_exists('tellafriend_civicrm_config')) {
```
```suggestion
    if (!\Civi\Api4\Extension::get(FALSE)->addWhere('file', '=', 'tellafriend')->addWhere('status', '=', 'installed')->execute()->selectRowCount()->count()) {
```

herbdool commented on CRM/Event/Form/ManageEvent/TabHeader.php:
```diff
@@ -85,7 +85,10 @@ public static function process(&$form) {
 
-    $tabs['friend'] = ['title' => ts('Tell a Friend')] + $default;
+    // @todo Move to hook_civicrm_tabset()
+    if (function_exists('tellafriend_civicrm_config')) {
```
```suggestion
    if (\Civi\Api4\Extension::get(FALSE)->addWhere('file', '=', 'tellafriend')->addWhere('status', '=', 'installed')->execute()->selectRowCount()->count()) {
```

herbdool commented on CRM/Event/Form/Registration/ThankYou.php:
```diff
@@ -152,29 +152,33 @@ public function buildQuickForm() {
-        );
+    $friendURL = NULL;
+
+    if (function_exists('tellafriend_civicrm_config')) {
```
```suggestion
    if (\Civi\Api4\Extension::get(FALSE)->addWhere('file', '=', 'tellafriend')->addWhere('status', '=', 'installed')->execute()->selectRowCount()->count()) {
```

herbdool commented on CRM/Event/Page/ManageEvent.php:
```diff
@@ -178,12 +178,14 @@ public static function &tabs() {
-          'url' => 'civicrm/event/manage/friend',
-          'field' => 'friend',
-        ];
+      if (!function_exists('tellafriend_civicrm_config')) {
```
```suggestion
      if (!\Civi\Api4\Extension::get(FALSE)->addWhere('file', '=', 'tellafriend')->addWhere('status', '=', 'installed')->execute()->selectRowCount()->count()) {
```

herbdool commented on CRM/Event/Page/ManageEvent.php:
```diff
@@ -405,7 +407,9 @@ public function browse() {
         $defaults['location'] = CRM_Core_BAO_Location::getValues($params, TRUE);
 
-        $manageEvent[$dao->id]['friend'] = CRM_Friend_BAO_Friend::getValues($params);
+        if (function_exists('tellafriend_civicrm_config')) {
```
```suggestion
        if (\Civi\Api4\Extension::get(FALSE)->addWhere('file', '=', 'tellafriend')->addWhere('status', '=', 'installed')->execute()->selectRowCount()->count()) {
```

mattwire commented:
@mlutfy merge conflict. Also one issue reported by @herbdool. If you can resolve that issue let's get this merged. I'm well aware how much work this kind of thing requires - since I did it for recaptcha and some others.

mlutfy commented:
Excellent - thank you for testing!

I rebased and fixed the missing xml file for the route, which was causing a 404. In my testing, I also noticed that tellafriend requires either a valid contribution/registration, or users must be logged-in (it statusBounces otherwise, which confused me at first). I tweaked the error message.

mattwire commented:
Tests passing / no conflicts. I've done some basic testing and am relying on @herbdool for the more detailed testing. It will be much easier to follow up with additional PRs for improvements / additional fixes after this one is merged.
